File: ios/ProviiWallet/ProviiWallet/Features/Help/Glossary.swift
// SPDX-License-Identifier: BUSL-1.1
// Copyright (c) 2024-2026 Maelstrom AI Pty Ltd ATF Maelstrom AI Holding Trust

import Foundation

/// Glossary data model and singleton providing searchable, categorised definitions of cryptographic,
/// privacy, credential, verification, and technical terms used throughout Provii Wallet. Includes
/// pronunciation guides for screen reader accessibility and related-term cross-references.

enum GlossaryCategory: String, CaseIterable {
    case cryptography = "Cryptography"
    case privacy = "Privacy"
    case credentials = "Credentials"
    case verification = "Verification"
    case technical = "Technical"
}

struct GlossaryEntry {
    let term: String
    let category: GlossaryCategory
    let definition: String
    let pronunciation: String?
    let relatedTerms: [String]

    init(term: String, category: GlossaryCategory, definition: String, pronunciation: String? = nil, relatedTerms: [String] = []) {
        self.term = term
        self.category = category
        self.definition = definition
        self.pronunciation = pronunciation
        self.relatedTerms = relatedTerms
    }
}

class Glossary {
    static let shared = Glossary()

    private init() {}

    func getAllEntries() -> [GlossaryEntry] {
        return entries
    }

    func entriesByCategory(_ category: GlossaryCategory) -> [GlossaryEntry] {
        return entries.filter { $0.category == category }
    }

    func findEntry(term: String) -> GlossaryEntry? {
        return entries.first { $0.term.caseInsensitiveCompare(term) == .orderedSame }
    }

    func search(query: String) -> [GlossaryEntry] {
        let lowered = query.lowercased()
        return entries.filter {
            $0.term.lowercased().contains(lowered) ||
            $0.definition.lowercased().contains(lowered)
        }
    }

    /// Get pronunciation-friendly term for accessibility
    func getPronunciationFriendlyTerm(_ term: String) -> String {
        if let entry = entries.first(where: { $0.term == term }),
           let pronunciation = entry.pronunciation {
            return pronunciation
        }
        return term
    }

    // MARK: - Entries (matching Android's 16 terms + iOS accessibility extras)

    private var entries: [GlossaryEntry] {
        return [
            // Cryptography
            GlossaryEntry(
                term: NSLocalizedString("glossary_term_zero_knowledge_proof", comment: "Glossary term: Zero Knowledge Proof"),
                category: .cryptography,
                definition: NSLocalizedString("glossary_def_zero_knowledge_proof", comment: "Glossary definition for Zero Knowledge Proof"),
                relatedTerms: ["Pedersen Commitment", "Range Proof", "Verifier", "Privacy Preserving"]
            ),
            GlossaryEntry(
                term: NSLocalizedString("glossary_term_pedersen_commitment", comment: "Glossary term: Pedersen Commitment"),
                category: .cryptography,
                definition: NSLocalizedString("glossary_def_pedersen_commitment", comment: "Glossary definition for Pedersen Commitment"),
                relatedTerms: ["Zero Knowledge Proof", "Cryptographic Commitment", "Homomorphic Encryption", "Range Proof"]
            ),
            GlossaryEntry(
                term: NSLocalizedString("glossary_term_range_proof", comment: "Glossary term: Range Proof"),
                category: .cryptography,
                definition: NSLocalizedString("glossary_def_range_proof", comment: "Glossary definition for Range Proof"),
                relatedTerms: ["Zero Knowledge Proof", "Pedersen Commitment", "Verification", "Data Minimisation"]
            ),
            GlossaryEntry(
                term: NSLocalizedString("glossary_term_cryptographic_commitment", comment: "Glossary term: Cryptographic Commitment"),
                category: .cryptography,
                definition: NSLocalizedString("glossary_def_cryptographic_commitment", comment: "Glossary definition for Cryptographic Commitment"),
                relatedTerms: ["Pedersen Commitment", "Zero Knowledge Proof", "Credential", "Verifier"]
            ),
            GlossaryEntry(
                term: NSLocalizedString("glossary_term_digital_signature", comment: "Glossary term: Digital Signature"),
                category: .cryptography,
                definition: NSLocalizedString("glossary_def_digital_signature", comment: "Glossary definition for Digital Signature"),
                relatedTerms: ["Issuer", "Credential", "Public Key Infrastructure", "Trusted Authority"]
            ),
            GlossaryEntry(
                term: NSLocalizedString("glossary_term_proving_key", comment: "Glossary term: Proving Key"),
                category: .cryptography,
                definition: NSLocalizedString("glossary_def_proving_key", comment: "Glossary definition for Proving Key"),
                relatedTerms: ["Zero Knowledge Proof", "Verification Key", "Cryptographic Parameters", "Digital Signature"]
            ),
            GlossaryEntry(
                term: NSLocalizedString("glossary_term_homomorphic_encryption", comment: "Glossary term: Homomorphic Encryption"),
                category: .cryptography,
                definition: NSLocalizedString("glossary_def_homomorphic_encryption", comment: "Glossary definition for Homomorphic Encryption"),
                relatedTerms: ["Pedersen Commitment", "Zero Knowledge Proof", "Range Proof", "Privacy Preserving"]
            ),

            // Privacy
            GlossaryEntry(
                term: NSLocalizedString("glossary_term_privacy_preserving", comment: "Glossary term: Privacy Preserving"),
                category: .privacy,
                definition: NSLocalizedString("glossary_def_privacy_preserving", comment: "Glossary definition for Privacy Preserving"),
                relatedTerms: ["Zero Knowledge Proof", "Privacy by Design", "Data Minimisation", "Credential"]
            ),

            // Credentials
            GlossaryEntry(
                term: NSLocalizedString("glossary_term_credential", comment: "Glossary term: Credential"),
                category: .credentials,
                definition: NSLocalizedString("glossary_def_credential", comment: "Glossary definition for Credential")
            ),
            GlossaryEntry(
                term: NSLocalizedString("glossary_term_issuer", comment: "Glossary term: Issuer"),
                category: .credentials,
                definition: NSLocalizedString("glossary_def_issuer", comment: "Glossary definition for Issuer")
            ),
            GlossaryEntry(
                term: NSLocalizedString("glossary_term_trusted_authority", comment: "Glossary term: Trusted Authority"),
                category: .credentials,
                definition: NSLocalizedString("glossary_def_trusted_authority", comment: "Glossary definition for Trusted Authority"),
                relatedTerms: ["Issuer", "Credential", "Digital Signature", "Public Key Infrastructure"]
            ),

            // Verification
            GlossaryEntry(
                term: NSLocalizedString("glossary_term_verifier", comment: "Glossary term: Verifier"),
                category: .verification,
                definition: NSLocalizedString("glossary_def_verifier", comment: "Glossary definition for Verifier")
            ),
            GlossaryEntry(
                term: NSLocalizedString("glossary_term_challenge", comment: "Glossary term: Challenge"),
                category: .verification,
                definition: NSLocalizedString("glossary_def_challenge", comment: "Glossary definition for Challenge"),
                relatedTerms: ["Verifier", "Zero Knowledge Proof", "Session", "Proving Key"]
            ),
            GlossaryEntry(
                term: NSLocalizedString("glossary_term_verification_key", comment: "Glossary term: Verification Key"),
                category: .verification,
                definition: NSLocalizedString("glossary_def_verification_key", comment: "Glossary definition for Verification Key"),
                relatedTerms: ["Proving Key", "Verifier", "Zero Knowledge Proof", "Cryptographic Parameters"]
            ),
            GlossaryEntry(
                term: NSLocalizedString("glossary_term_session", comment: "Glossary term: Session"),
                category: .verification,
                definition: NSLocalizedString("glossary_def_session", comment: "Glossary definition for Session"),
                relatedTerms: ["Challenge", "Verifier", "Timeout", "Zero Knowledge Proof"]
            ),

            // Technical
            GlossaryEntry(
                term: NSLocalizedString("glossary_term_qr_code", comment: "Glossary term: QR Code"),
                category: .technical,
                definition: NSLocalizedString("glossary_def_qr_code", comment: "Glossary definition for QR Code"),
                pronunciation: "Q R Code"
            ),
            GlossaryEntry(
                term: NSLocalizedString("glossary_term_biometric_authentication", comment: "Glossary term: Biometric Authentication"),
                category: .technical,
                definition: NSLocalizedString("glossary_def_biometric_authentication", comment: "Glossary definition for Biometric Authentication")
            ),
            GlossaryEntry(
                term: NSLocalizedString("glossary_term_api", comment: "Glossary term: API"),
                category: .technical,
                definition: NSLocalizedString("glossary_def_api", comment: "Glossary definition for API"),
                pronunciation: "A P I"
            ),
            GlossaryEntry(
                term: NSLocalizedString("glossary_term_url", comment: "Glossary term: URL"),
                category: .technical,
                definition: NSLocalizedString("glossary_def_url", comment: "Glossary definition for URL"),
                pronunciation: "U R L"
            ),
            GlossaryEntry(
                term: NSLocalizedString("glossary_term_id", comment: "Glossary term: ID"),
                category: .technical,
                definition: NSLocalizedString("glossary_def_id", comment: "Glossary definition for ID"),
                pronunciation: "I D"
            ),
            GlossaryEntry(
                term: NSLocalizedString("glossary_term_mdl", comment: "Glossary term: mDL"),
                category: .technical,
                definition: NSLocalizedString("glossary_def_mdl", comment: "Glossary definition for mDL"),
                pronunciation: "M D L"
            ),
            GlossaryEntry(
                term: NSLocalizedString("glossary_term_pin", comment: "Glossary term: PIN"),
                category: .technical,
                definition: NSLocalizedString("glossary_def_pin", comment: "Glossary definition for PIN"),
                pronunciation: "P I N"
            ),
            GlossaryEntry(
                term: "Provii",
                category: .technical,
                definition: NSLocalizedString("glossary_def_provii", comment: "Glossary definition for Provii"),
                pronunciation: "par lee"
            )
        ]
    }
}

File: android/app/src/main/java/app/provii/wallet/ui/accessibility/Glossary.kt
// SPDX-License-Identifier: BUSL-1.1
// Copyright (c) 2024-2026 Maelstrom AI Pty Ltd ATF Maelstrom AI Holding Trust
package app.provii.wallet.ui.accessibility

import androidx.annotation.StringRes
import app.provii.wallet.R

/**
 * In-app technical glossary satisfying WCAG 2.2 AAA criterion 3.1.3 (Unusual Words).
 * Contains 16 entries organised into five categories: cryptography, privacy, credentials,
 * verification, and technical. Each entry carries a short definition for tooltips and a
 * full definition for the dedicated glossary screen, plus related-term cross-references.
 *
 * All user-visible strings are held as @StringRes resource IDs. Composables resolve them
 * via stringResource() so translations flow through the standard Android localisation pipeline.
 */

enum class GlossaryCategory {
    CRYPTOGRAPHY,
    PRIVACY,
    CREDENTIALS,
    VERIFICATION,
    TECHNICAL,
}

data class GlossaryEntry(
    @StringRes val termRes: Int,
    val category: GlossaryCategory,
    @StringRes val shortDefinitionRes: Int,
    @StringRes val fullDefinitionRes: Int,
    val relatedTerms: List<String> = emptyList(),
)

object Glossary {
    private val entries =
        listOf(
            GlossaryEntry(
                termRes = R.string.glossary_term_zero_knowledge_proof,
                category = GlossaryCategory.CRYPTOGRAPHY,
                shortDefinitionRes = R.string.glossary_short_zero_knowledge_proof,
                fullDefinitionRes = R.string.glossary_full_zero_knowledge_proof,
                relatedTerms = listOf("Pedersen Commitment", "Range Proof", "Verifier", "Privacy Preserving"),
            ),
            GlossaryEntry(
                termRes = R.string.glossary_term_credential,
                category = GlossaryCategory.CREDENTIALS,
                shortDefinitionRes = R.string.glossary_short_credential,
                fullDefinitionRes = R.string.glossary_full_credential,
                relatedTerms = listOf("Issuer", "Digital Signature", "Cryptographic Commitment", "Attestation"),
            ),
            GlossaryEntry(
                termRes = R.string.glossary_term_issuer,
                category = GlossaryCategory.CREDENTIALS,
                shortDefinitionRes = R.string.glossary_short_issuer,
                fullDefinitionRes = R.string.glossary_full_issuer,
                relatedTerms = listOf("Credential", "Digital Signature", "Trusted Authority", "Attestation"),
            ),
            GlossaryEntry(
                termRes = R.string.glossary_term_verifier,
                category = GlossaryCategory.VERIFICATION,
                shortDefinitionRes = R.string.glossary_short_verifier,
                fullDefinitionRes = R.string.glossary_full_verifier,
                relatedTerms = listOf("Zero Knowledge Proof", "Range Proof", "Challenge", "Session"),
            ),
            GlossaryEntry(
                termRes = R.string.glossary_term_qr_code,
                category = GlossaryCategory.TECHNICAL,
                shortDefinitionRes = R.string.glossary_short_qr_code,
                fullDefinitionRes = R.string.glossary_full_qr_code,
                relatedTerms = listOf("Manual Code Entry", "Scanning", "Alternative Input", "Deep Link"),
            ),
            GlossaryEntry(
                termRes = R.string.glossary_term_pedersen_commitment,
                category = GlossaryCategory.CRYPTOGRAPHY,
                shortDefinitionRes = R.string.glossary_short_pedersen_commitment,
                fullDefinitionRes = R.string.glossary_full_pedersen_commitment,
                relatedTerms = listOf("Zero Knowledge Proof", "Cryptographic Commitment", "Homomorphic Encryption", "Range Proof"),
            ),
            GlossaryEntry(
                termRes = R.string.glossary_term_range_proof,
                category = GlossaryCategory.CRYPTOGRAPHY,
                shortDefinitionRes = R.string.glossary_short_range_proof,
                fullDefinitionRes = R.string.glossary_full_range_proof,
                relatedTerms = listOf("Zero Knowledge Proof", "Pedersen Commitment", "Verification", "Age Threshold"),
            ),
            GlossaryEntry(
                termRes = R.string.glossary_term_challenge,
                category = GlossaryCategory.VERIFICATION,
                shortDefinitionRes = R.string.glossary_short_challenge,
                fullDefinitionRes = R.string.glossary_full_challenge,
                relatedTerms = listOf("Verifier", "Zero Knowledge Proof", "Session", "Nonce"),
            ),
            GlossaryEntry(
                termRes = R.string.glossary_term_cryptographic_commitment,
                category = GlossaryCategory.CRYPTOGRAPHY,
                shortDefinitionRes = R.string.glossary_short_cryptographic_commitment,
                fullDefinitionRes = R.string.glossary_full_cryptographic_commitment,
                relatedTerms = listOf("Pedersen Commitment", "Zero Knowledge Proof", "Credential", "Binding Property"),
            ),
            GlossaryEntry(
                termRes = R.string.glossary_term_digital_signature,
                category = GlossaryCategory.CRYPTOGRAPHY,
                shortDefinitionRes = R.string.glossary_short_digital_signature,
                fullDefinitionRes = R.string.glossary_full_digital_signature,
                relatedTerms = listOf("Issuer", "Credential", "Public Key Infrastructure", "Ed25519"),
            ),
            GlossaryEntry(
                termRes = R.string.glossary_term_privacy_preserving,
                category = GlossaryCategory.PRIVACY,
                shortDefinitionRes = R.string.glossary_short_privacy_preserving,
                fullDefinitionRes = R.string.glossary_full_privacy_preserving,
                relatedTerms = listOf("Zero Knowledge Proof", "Privacy by Design", "Data Minimisation", "Selective Disclosure"),
            ),
            GlossaryEntry(
                termRes = R.string.glossary_term_proving_key,
                category = GlossaryCategory.CRYPTOGRAPHY,
                shortDefinitionRes = R.string.glossary_short_proving_key,
                fullDefinitionRes = R.string.glossary_full_proving_key,
                relatedTerms = listOf("Zero Knowledge Proof", "Verification Key", "Cryptographic Parameters", "Trusted Setup"),
            ),
            GlossaryEntry(
                termRes = R.string.glossary_term_verification_key,
                category = GlossaryCategory.VERIFICATION,
                shortDefinitionRes = R.string.glossary_short_verification_key,
                fullDefinitionRes = R.string.glossary_full_verification_key,
                relatedTerms = listOf("Proving Key", "Verifier", "Zero Knowledge Proof", "Public Parameters"),
            ),
            GlossaryEntry(
                termRes = R.string.glossary_term_session,
                category = GlossaryCategory.VERIFICATION,
                shortDefinitionRes = R.string.glossary_short_session,
                fullDefinitionRes = R.string.glossary_full_session,
                relatedTerms = listOf("Challenge", "Verifier", "Timeout", "Nonce"),
            ),
            GlossaryEntry(
                termRes = R.string.glossary_term_homomorphic_encryption,
                category = GlossaryCategory.CRYPTOGRAPHY,
                shortDefinitionRes = R.string.glossary_short_homomorphic_encryption,
                fullDefinitionRes = R.string.glossary_full_homomorphic_encryption,
                relatedTerms = listOf("Pedersen Commitment", "Zero Knowledge Proof", "Range Proof", "Additive Property"),
            ),
            GlossaryEntry(
                termRes = R.string.glossary_term_trusted_authority,
                category = GlossaryCategory.CREDENTIALS,
                shortDefinitionRes = R.string.glossary_short_trusted_authority,
                fullDefinitionRes = R.string.glossary_full_trusted_authority,
                relatedTerms = listOf("Issuer", "Credential", "Digital Signature", "Certificate Authority"),
            ),
        )

    fun allEntries(): List<GlossaryEntry> = entries

    fun entriesByCategory(category: GlossaryCategory): List<GlossaryEntry> =
        entries.filter { it.category == category }

    fun getAllCategories(): List<GlossaryCategory> = GlossaryCategory.entries.toList()
}
